// @webzaytsev/yookassa-ts-sdk/dist/types/general.types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CurrencyEnum = exports.LocaleEnum = void 0;
/**
 * Supported locales for YooKassa interface
 * @see https://yookassa.ru/developers/api#create_payment_confirmation_locale
 */
var LocaleEnum;
(function (LocaleEnum) {
    /** Russian */
    LocaleEnum["ru_RU"] = "ru_RU";
    /** English */
    LocaleEnum["en_US"] = "en_US";
})(LocaleEnum || (exports.LocaleEnum = LocaleEnum = {}));
/**
 * Supported currencies in ISO-4217 format.
 * @see https://yookassa.ru/developers/api#payment_object_amount_currency
 */
var CurrencyEnum;
(function (CurrencyEnum) {
    /** Российский рубль */
    CurrencyEnum["RUB"] = "RUB";
    /** Евро */
    CurrencyEnum["EUR"] = "EUR";
    /** Доллар США */
    CurrencyEnum["USD"] = "USD";
    /** Казахстанский тенге */
    CurrencyEnum["KZT"] = "KZT";
    /** Белорусский рубль */
    CurrencyEnum["BYN"] = "BYN";
    /** Украинская гривна */
    CurrencyEnum["UAH"] = "UAH";
    /** Узбекский сум */
    CurrencyEnum["UZS"] = "UZS";
})(CurrencyEnum || (exports.CurrencyEnum = CurrencyEnum = {}));
