// @webzaytsev/yookassa-ts-sdk/dist/client/sdk.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.YooKassaSdk = void 0;
exports.YooKassa = YooKassa;
exports.clearYooKassaCache = clearYooKassaCache;
const connector_1 = require("./connector");
/**
 * YooKassa SDK client for payment processing.
 *
 * ## Features
 *
 * - **Automatic retries** with exponential backoff on network errors and 5xx responses
 * - **Built-in rate limiting** via `maxRPS` configuration
 * - **Idempotency** — auto-generated or custom keys for safe retries
 * - **Instance caching** by `shop_id` for connection reuse
 *
 * ## Usage
 *
 * ```typescript
 * import { YooKassa } from '@webzaytsev/yookassa-ts-sdk';
 *
 * const sdk = YooKassa({
 *     shop_id: 'your_shop_id',
 *     secret_key: 'your_secret_key',
 * });
 *
 * // Create payment with auto-generated idempotency key
 * const payment = await sdk.payments.create({
 *     amount: { value: '100.00', currency: 'RUB' },
 *     confirmation: { type: 'redirect', return_url: 'https://example.com' },
 * });
 *
 * // Create payment with custom idempotency key (recommended for retries)
 * const payment = await sdk.payments.create(paymentData, `order-${orderId}`);
 * ```
 *
 * @see https://yookassa.ru/developers/api
 */
class YooKassaSdk extends connector_1.Connector {
    constructor() {
        super(...arguments);
        // ========== Payments ========== //
        this.getPaymentById = (id) => __awaiter(this, void 0, void 0, function* () {
            return this.request({
                method: 'GET',
                endpoint: `/payments/${id}`,
            });
        });
        /** ****Получить список платежей****
         *
         * Запрос позволяет получить список платежей, отфильтрованный по заданным критериям. [Подробнее о работе со списками](https://yookassa.ru/developers/using-api/lists)
         */
        this.getPaymentList = (filter) => __awaiter(this, void 0, void 0, function* () {
            return this.fetchList('/payments', filter);
        });
        /** Создать платеж
         * @param newPayment - данные для создания платежа
         * @param idempotenceKey - ключ идемпотентности (опционально, генерируется автоматически)
         */
        this.createPayment = (newPayment, idempotenceKey) => __awaiter(this, void 0, void 0, function* () {
            return this.request({
                method: 'POST',
                endpoint: '/payments',
                data: newPayment,
                requestId: idempotenceKey,
            });
        });
        /** Подтвердить платеж по идентификатору
         * @param paymentId - идентификатор платежа
         * @param payload - данные для подтверждения (сумма, чек, данные авиабилетов, распределение)
         * @param idempotenceKey - ключ идемпотентности (опционально)
         */
        this.capturePaymentById = (paymentId, payload, idempotenceKey) => __awaiter(this, void 0, void 0, function* () {
            return this.request({
                method: 'POST',
                endpoint: `/payments/${paymentId}/capture`,
                data: payload !== null && payload !== void 0 ? payload : {},
                requestId: idempotenceKey,
            });
        });
        /** Отменить платеж по идентификатору
         * @param paymentId - идентификатор платежа
         * @param idempotenceKey - ключ идемпотентности (опционально)
         */
        this.cancelPaymentById = (paymentId, idempotenceKey) => __awaiter(this, void 0, void 0, function* () {
            return this.request({
                method: 'POST',
                endpoint: `/payments/${paymentId}/cancel`,
                data: {},
                requestId: idempotenceKey,
            });
        });
        // ========== Payments end========== //
        // ========== Refunds ========== //
        this.getRefundById = (refundId) => __awaiter(this, void 0, void 0, function* () {
            return this.request({
                method: 'GET',
                endpoint: `/refunds/${refundId}`,
            });
        });
        this.getRefundList = (filter) => __awaiter(this, void 0, void 0, function* () {
            return this.fetchList('/refunds', filter);
        });
        /** Создать возврат
         * @param newRefund - данные для создания возврата
         * @param idempotenceKey - ключ идемпотентности (опционально)
         */
        this.createRefund = (newRefund, idempotenceKey) => __awaiter(this, void 0, void 0, function* () {
            return this.request({
                method: 'POST',
                endpoint: '/refunds',
                data: newRefund,
                requestId: idempotenceKey,
            });
        });
        // ========== Refunds end ========== //
        // ========== Receipts ========== //
        this.getReceiptById = (receiptId) => __awaiter(this, void 0, void 0, function* () {
            return this.request({
                method: 'GET',
                endpoint: `/receipts/${receiptId}`,
            });
        });
        /** Создать чек
         * @param newReceipt - данные для создания чека
         * @param idempotenceKey - ключ идемпотентности (опционально)
         */
        this.createReceipt = (newReceipt, idempotenceKey) => __awaiter(this, void 0, void 0, function* () {
            return this.request({
                method: 'POST',
                endpoint: '/receipts',
                data: newReceipt,
                requestId: idempotenceKey,
            });
        });
        this.getReceiptList = (filter) => __awaiter(this, void 0, void 0, function* () {
            return this.fetchList('/receipts', filter);
        });
        // ========== Receipts end ========== //
        // ========== Webhooks ========== //
        /** Создать вебхук (требуется OAuth-токен)
         * @param webhook - данные для создания вебхука
         * @param idempotenceKey - ключ идемпотентности (опционально)
         */
        this.createWebhook = (webhook, idempotenceKey) => __awaiter(this, void 0, void 0, function* () {
            return this.request({
                method: 'POST',
                endpoint: '/webhooks',
                data: webhook,
                requestId: idempotenceKey,
                useOAuth: true,
            });
        });
        /** Получить список вебхуков (требуется OAuth-токен) */
        this.getWebhookList = () => __awaiter(this, void 0, void 0, function* () {
            const response = yield this.request({
                method: 'GET',
                endpoint: '/webhooks',
                useOAuth: true,
            });
            return response.items;
        });
        /** Удалить вебхук по идентификатору (требуется OAuth-токен)
         * @param webhookId - идентификатор вебхука
         */
        this.deleteWebhook = (webhookId) => __awaiter(this, void 0, void 0, function* () {
            yield this.request({
                method: 'DELETE',
                endpoint: `/webhooks/${webhookId}`,
                useOAuth: true,
            });
        });
        // ========== Webhooks end ========== //
        // ========== Payouts ========== //
        this.getPayoutById = (id) => __awaiter(this, void 0, void 0, function* () {
            return this.request({
                method: 'GET',
                endpoint: `/payouts/${id}`,
            });
        });
        this.getPayoutList = (filter) => __awaiter(this, void 0, void 0, function* () {
            return this.fetchList('/payouts', filter);
        });
        this.searchPayouts = (filter) => __awaiter(this, void 0, void 0, function* () {
            return this.fetchList('/payouts/search', filter);
        });
        this.createPayout = (newPayout, idempotenceKey) => __awaiter(this, void 0, void 0, function* () {
            return this.request({
                method: 'POST',
                endpoint: '/payouts',
                data: newPayout,
                requestId: idempotenceKey,
            });
        });
        // ========== Payouts end ========== //
        // ========== Sbp banks ========== //
        this.getSbpBanks = () => __awaiter(this, void 0, void 0, function* () {
            const response = yield this.request({
                method: 'GET',
                endpoint: '/sbp_banks',
            });
            return response.items;
        });
        // ========== Sbp banks end ========== //
        // ========== Saved payment methods ========== //
        this.createSavedPaymentMethod = (data, idempotenceKey) => __awaiter(this, void 0, void 0, function* () {
            return this.request({
                method: 'POST',
                endpoint: '/payment_methods',
                data,
                requestId: idempotenceKey,
            });
        });
        this.getSavedPaymentMethodById = (paymentMethodId) => __awaiter(this, void 0, void 0, function* () {
            return this.request({
                method: 'GET',
                endpoint: `/payment_methods/${paymentMethodId}`,
            });
        });
        // ========== Saved payment methods end ========== //
        // ========== Personal data ========== //
        this.createPersonalDataRecord = (data, idempotenceKey) => __awaiter(this, void 0, void 0, function* () {
            return this.request({
                method: 'POST',
                endpoint: '/personal_data',
                data,
                requestId: idempotenceKey,
            });
        });
        this.getPersonalDataById = (personalDataId) => __awaiter(this, void 0, void 0, function* () {
            return this.request({
                method: 'GET',
                endpoint: `/personal_data/${personalDataId}`,
            });
        });
        // ========== Personal data end ========== //
        // ========== Deals ========== //
        this.createDeal = (data, idempotenceKey) => __awaiter(this, void 0, void 0, function* () {
            return this.request({
                method: 'POST',
                endpoint: '/deals',
                data,
                requestId: idempotenceKey,
            });
        });
        this.getDealList = (filter) => __awaiter(this, void 0, void 0, function* () {
            return this.fetchList('/deals', filter);
        });
        this.getDealById = (dealId) => __awaiter(this, void 0, void 0, function* () {
            return this.request({
                method: 'GET',
                endpoint: `/deals/${dealId}`,
            });
        });
        // ========== Deals end ========== //
        // ========== Invoices ========== //
        this.createInvoice = (data, idempotenceKey) => __awaiter(this, void 0, void 0, function* () {
            return this.request({
                method: 'POST',
                endpoint: '/invoices',
                data,
                requestId: idempotenceKey,
            });
        });
        this.getInvoiceById = (invoiceId) => __awaiter(this, void 0, void 0, function* () {
            return this.request({
                method: 'GET',
                endpoint: `/invoices/${invoiceId}`,
            });
        });
        // ========== Invoices end ========== //
        // ========== Shop ========== //
        /** Получить информацию о магазине (требуется OAuth-токен) */
        this.getShopInfo = () => __awaiter(this, void 0, void 0, function* () {
            return this.request({
                method: 'GET',
                endpoint: '/me',
                useOAuth: true,
            });
        });
        // ========== Shop end ========== //
        /** Методы для работы с платежами */
        this.payments = {
            /**
             * Creates a new payment.
             *
             * ## Idempotency
             *
             * The `idempotenceKey` parameter ensures safe retries in distributed systems.
             * If not provided, SDK generates a UUID automatically.
             *
             * **Recommended**: Use order ID or business identifier as idempotency key
             * to prevent duplicate payments on retries.
             *
             * @param payment - Payment data (amount, confirmation, etc.)
             * @param idempotenceKey - Unique key for request deduplication. Use order ID for safe retries.
             *
             * @example
             * ```typescript
             * // Auto-generated idempotency key (for single attempts)
             * const payment = await sdk.payments.create({
             *     amount: { value: '100.00', currency: 'RUB' },
             *     confirmation: { type: 'redirect', return_url: 'https://example.com' },
             * });
             *
             * // Custom idempotency key (recommended for retries)
             * const payment = await sdk.payments.create(paymentData, `order-${orderId}`);
             *
             * // Safe retry pattern for distributed systems
             * async function createPaymentWithRetry(orderId: string, data: CreatePaymentRequest) {
             *     const key = `payment-${orderId}`;
             *     try {
             *         return await sdk.payments.create(data, key);
             *     } catch (error) {
             *         // Same key = same result, no duplicate charge
             *         return await sdk.payments.create(data, key);
             *     }
             * }
             * ```
             *
             * @see https://yookassa.ru/developers/api#create_payment
             */
            create: this.createPayment,
            /**
             * ****Информация о платеже****
             *
             * Запрос позволяет получить информацию о текущем состоянии платежа по его уникальному идентификатору.
             *
             * [Документация](https://yookassa.ru/developers/api#get_payment)
             */
            load: this.getPaymentById,
            /**
             * ****Список платежей****
             *
             * Запрос позволяет получить список платежей, отфильтрованный по заданным критериям.
             *
             * Retrieves a list of payments filtered by the specified criteria.
             *
             * @param {Omit<GetPaymentListFilter, 'cursor'>} params - The parameters for filtering the payments.
             * @return {Promise<Payments.IPayment[]>} A promise that resolves to the list of payments.
             *
             * [API Documentation](https://yookassa.ru/developers/api#get_payment)
             *
             * [Working with lists documentation](https://yookassa.ru/developers/using-api/lists)
             *
             * **PaymentListParams**
             *
             * These are the parameters that can be used to filter the payments.
             *
             * **created_at**
             * - `gte` (optional): Filter by the creation time. The time should be greater than or equal to the specified value. Format: ISO 8601. Example: `created_at.gte=2018-07-18T10:51:18.139Z`
             * - `gt` (optional): Filter by the creation time. The time should be greater than the specified value. Format: ISO 8601. Example: `created_at.gt=2018-07-18T10:51:18.139Z`
             * - `lte` (optional): Filter by the creation time. The time should be less than or equal to the specified value. Format: ISO 8601. Example: `created_at.lte=2018-07-18T10:51:18.139Z`
             * - `lt` (optional): Filter by the creation time. The time should be less than the specified value. Format: ISO 8601. Example: `created_at.lt=2018-07-18T10:51:18.139Z`
             *
             * **captured_at**
             * - `gte` (optional): Filter by the time of payment confirmation. The time should be greater than or equal to the specified value. Format: ISO 8601. Example: `captured_at.gte=2018-07-18T10:51:18.139Z`
             * - `gt` (optional): Filter by the time of payment confirmation. The time should be greater than the specified value. Format: ISO 8601. Example: `captured_at.gt=2018-07-18T10:51:18.139Z`
             * - `lte` (optional): Filter by the time of payment confirmation. The time should be less than or equal to the specified value. Format: ISO 8601. Example: `captured_at.lte=2018-07-18T10:51:18.139Z`
             * - `lt` (optional): Filter by the time of payment confirmation. The time should be less than the specified value. Format: ISO 8601. Example: `captured_at.lt=2018-07-18T10:51:18.139Z`
             *
             * **payment_method**
             * - `string` (optional): Filter by payment method code. Example: `payment_method=bank_card`
             *
             * **status**
             * - `string` (optional): Filter by payment status. Example: `status=succeeded`
             *
             * **limit**
             * - `number` (optional): The number of objects returned in the response. Possible values: from 1 to 100. Example: `limit=50`
             * - Default value: `10`
             *
             * **cursor**
             * - `string` (optional): Used to retrieve the next fragment of the list. Example: `cursor=37a5c87d-3984-51e8-a7f3-8de646d39ec15`
             * - Used as an indicator to retrieve the next fragment of the list. This should be used if there are more objects in the list than the number specified in the limit parameter. An example of how to use it is provided in the "Working with lists" documentation.
             *
             * **PaymentList**
             *
             * This is the response structure for the list of payments.
             *
             * **next_cursor**
             * - `string` (optional): Used to retrieve the next fragment of the list.
             *
             * **payments**
             * - `Payments.Payment[]` (optional): The list of payments.
             */
            list: this.getPaymentList,
            /**
             * ****Подтверждение платежа***
             *
             * Подтверждает вашу готовность принять платеж.
             * После подтверждения платеж перейдет в статус `succeeded`.
             * Это значит, что вы можете выдать товар или оказать услугу пользователю.
             * Подтвердить можно только платеж в статусе `waiting_for_capture`
             * и только в течение определенного времени (зависит от способа оплаты).
             * Если вы не подтвердите платеж в отведенное время, он автоматически
             * перейдет в статус `canceled`, и деньги вернутся пользователю.
             *
             * [Подробнее о подтверждении и отмене платежей](https://yookassa.ru/developers/payment-acceptance/getting-started/payment-process#capture-and-cancel)
             *
             * @param paymentId - идентификатор платежа
             * @param payload - данные для подтверждения (сумма, чек, авиабилеты, распределение)
             * @param idempotenceKey - ключ идемпотентности (опционально)
             */
            capture: this.capturePaymentById,
            /**
             * ****Отмена платежа****
             *
             * Отменяет платеж, находящийся в статусе `waiting_for_capture`.
             * Отмена платежа значит, что вы не готовы выдать пользователю товар или оказать услугу.
             * Как только вы отменяете платеж, мы начинаем возвращать деньги на счет плательщика.
             * Для платежей банковскими картами, из кошелька ЮMoney или через SberPay отмена происходит мгновенно.
             * Для остальных способов оплаты возврат может занимать до нескольких дней.
             *
             * [Подробнее о подтверждении и отмене платежей](https://yookassa.ru/developers/payment-acceptance/getting-started/payment-process#capture-and-cancel)
             *
             * @param paymentId - идентификатор платежа
             * @param idempotenceKey - ключ идемпотентности (опционально)
             */
            cancel: this.cancelPaymentById,
        };
        /** Методы для работы с возвратами */
        this.refunds = {
            /**
             * ****Создание возврата****
             *
             * Создает возврат успешного платежа на указанную сумму.
             * Платеж можно вернуть только в течение трех лет с момента его создания.
             * Комиссия ЮKassa за проведение платежа не возвращается.
             *
             * @param refund - данные возврата
             * @param idempotenceKey - ключ идемпотентности (опционально)
             * @see https://yookassa.ru/developers/api#create_refund
             */
            create: this.createRefund,
            /**
             * ****Информация о возврате****
             *
             * Запрос позволяет получить информацию о текущем состоянии возврата по его уникальному идентификатору.
             * @see https://yookassa.ru/developers/api#get_refund
             */
            load: this.getRefundById,
            /**
             * ****Список возвратов****
             *
             * Запрос позволяет получить список возвратов, отфильтрованный по заданным критериям.
             *
             * [Подробнее о работе со списками](https://yookassa.ru/developers/using-api/lists)
             * @see https://yookassa.ru/developers/api#get_refunds_list
             */
            list: this.getRefundList,
        };
        /**
         * ****Методы для работы с чеками****
         *
         * С помощью API можно получать информацию о чеках, для которых вы отправили данные через ЮKassa.
         * @see https://yookassa.ru/developers/api#receipt
         */
        this.receipts = {
            /**
             * ****Создание чека****
             *
             * Используйте этот запрос при оплате с соблюдением требований 54-ФЗ, чтобы создать чек зачета предоплаты.
             * Если вы работаете по сценарию [Сначала платеж, потом чек](https://yookassa.ru/developers/payment-acceptance/receipts/54fz/other-services/basics#receipt-after-payment),
             * в запросе также нужно передавать данные для формирования чека прихода и чека возврата прихода.
             *
             * @param receipt - данные чека
             * @param idempotenceKey - ключ идемпотентности (опционально)
             * @see https://yookassa.ru/developers/api#create_receipt
             */
            create: this.createReceipt,
            /**
             * ****Информация о чеке****
             *
             * Запрос позволяет получить информацию о текущем состоянии чека по его уникальному идентификатору.
             */
            load: this.getReceiptById,
            /**
             * ****Список чеков****
             *
             * Запрос позволяет получить список чеков, отфильтрованный по заданным критериям.
             * Можно запросить чеки по конкретному платежу, чеки по конкретному возврату или все чеки магазина.
             *
             * [Подробнее о работе со списками](https://yookassa.ru/developers/using-api/lists)
             * @see https://yookassa.ru/developers/api#get_receipts_list
             */
            list: this.getReceiptList,
        };
        /**
         * ****Методы для работы с выплатами****
         *
         * Позволяет создавать выплаты физическим лицам и получать информацию о них.
         * Доступно при использовании обычных выплат или в рамках Безопасной сделки.
         *
         * @see https://yookassa.ru/developers/payouts/overview
         */
        this.payouts = {
            /**
             * ****Создание выплаты****
             *
             * Создаёт выплату физическому лицу на указанное платёжное средство.
             *
             * @param payout - данные выплаты
             * @param idempotenceKey - ключ идемпотентности (опционально)
             * @see https://yookassa.ru/developers/api#create_payout
             */
            create: this.createPayout,
            /**
             * ****Информация о выплате****
             *
             * Возвращает информацию о текущем состоянии выплаты по её идентификатору.
             *
             * @see https://yookassa.ru/developers/api#get_payout
             */
            load: this.getPayoutById,
            /**
             * ****Список выплат****
             *
             * Возвращает список выплат, отфильтрованный по заданным критериям.
             * Поддерживает фильтрацию по `created_at` и `succeeded_at`.
             *
             * @see https://yookassa.ru/developers/api#get_payouts_list
             */
            list: this.getPayoutList,
            /**
             * ****Поиск выплат****
             *
             * Поиск по `metadata` и периоду `created_at` (только за последние 3 месяца).
             *
             * @see https://yookassa.ru/developers/api#search_payouts
             */
            search: this.searchPayouts,
        };
        /** Участники СБП (справочник для выплат через СБП). */
        this.sbpBanks = {
            /** @see https://yookassa.ru/developers/api#get_sbp_banks */
            list: this.getSbpBanks,
        };
        /** Создание и получение сохранённых способов оплаты (привязка карты и т.п.). */
        this.paymentMethods = {
            /**
             * @see https://yookassa.ru/developers/api#create_payment_method
             */
            create: this.createSavedPaymentMethod,
            /**
             * @see https://yookassa.ru/developers/api#get_payment_method
             */
            load: this.getSavedPaymentMethodById,
        };
        /** Персональные данные получателя (выплаты с проверкой / выписки). */
        this.personalData = {
            /**
             * @see https://yookassa.ru/developers/api#create_personal_data
             */
            create: this.createPersonalDataRecord,
            /**
             * @see https://yookassa.ru/developers/api#get_personal_data
             */
            load: this.getPersonalDataById,
        };
        /** Безопасная сделка. */
        this.deals = {
            /**
             * @see https://yookassa.ru/developers/api#create_deal
             */
            create: this.createDeal,
            /**
             * @see https://yookassa.ru/developers/api#get_deals_list
             */
            list: this.getDealList,
            /**
             * @see https://yookassa.ru/developers/api#get_deal
             */
            load: this.getDealById,
        };
        /** Выставленные счета (инвойсы). */
        this.invoices = {
            /**
             * @see https://yookassa.ru/developers/api#create_invoice
             */
            create: this.createInvoice,
            /**
             * @see https://yookassa.ru/developers/api#get_invoice
             */
            load: this.getInvoiceById,
        };
        /**
         * ****Методы для работы с вебхуками****
         *
         * Вебхуки позволяют получать уведомления о событиях в ЮKassa.
         * **Требуется OAuth-токен** — функционал доступен только в рамках партнёрской программы.
         *
         * @see https://yookassa.ru/developers/api#webhook
         */
        this.webhooks = {
            /**
             * ****Создание вебхука****
             *
             * Создаёт вебхук для получения уведомлений о событиях.
             *
             * @param webhook - данные вебхука (event, url)
             * @param idempotenceKey - ключ идемпотентности (опционально)
             * @see https://yookassa.ru/developers/api#create_webhook
             */
            create: this.createWebhook,
            /**
             * ****Список вебхуков****
             *
             * Возвращает список созданных вебхуков.
             * @see https://yookassa.ru/developers/api#get_webhook_list
             */
            list: this.getWebhookList,
            /**
             * ****Удаление вебхука****
             *
             * Удаляет вебхук по идентификатору.
             *
             * @param webhookId - идентификатор вебхука
             * @see https://yookassa.ru/developers/api#delete_webhook
             */
            delete: this.deleteWebhook,
        };
        /**
         * ****Информация о магазине****
         *
         * Позволяет получить информацию о подключённом магазине.
         * **Требуется OAuth-токен** — функционал доступен только в рамках партнёрской программы.
         *
         * @see https://yookassa.ru/developers/api#get_me
         */
        this.shop = {
            /**
             * ****Получить информацию о магазине****
             *
             * Возвращает информацию об аккаунте: идентификатор, статус, доступные способы оплаты и т.д.
             * @see https://yookassa.ru/developers/api#get_me
             */
            info: this.getShopInfo,
        };
    }
    /**
     * Generic method for fetching lists with automatic pagination
     */
    fetchList(endpoint, filter) {
        return __awaiter(this, void 0, void 0, function* () {
            const items = [];
            const params = (filter !== null && filter !== void 0 ? filter : {});
            let cursor;
            do {
                const response = yield this.request({
                    method: 'GET',
                    endpoint,
                    params,
                });
                items.push(...response.items);
                cursor = response.next_cursor;
                if (cursor) {
                    params.cursor = cursor;
                }
            } while (cursor);
            return items;
        });
    }
}
exports.YooKassaSdk = YooKassaSdk;
/** Кэш инстансов SDK по shop_id */
const clientCache = new Map();
/**
 * Creates or returns a cached YooKassaSdk instance.
 *
 * Instances are cached by `shop_id`, enabling connection reuse
 * and multi-store support in a single application.
 *
 * ## Instance Caching
 *
 * ```typescript
 * // Same shop_id = same instance (cached)
 * const sdk1 = YooKassa({ shop_id: '123', secret_key: 'key' });
 * const sdk2 = YooKassa({ shop_id: '123', secret_key: 'key' });
 * console.log(sdk1 === sdk2); // true
 *
 * // Different shops = different instances
 * const shopA = YooKassa({ shop_id: 'A', secret_key: 'keyA' });
 * const shopB = YooKassa({ shop_id: 'B', secret_key: 'keyB' });
 * ```
 *
 * ## Distributed Systems Configuration
 *
 * ```typescript
 * // For multiple server instances sharing API rate limits
 * const sdk = YooKassa({
 *     shop_id: process.env.YOOKASSA_SHOP_ID,
 *     secret_key: process.env.YOOKASSA_SECRET_KEY,
 *     maxRPS: 2,       // Low per-instance limit
 *     retries: 5,      // More retries for resilience
 *     timeout: 15000,  // Longer timeout
 * });
 * ```
 *
 * @param init - SDK configuration options
 * @param forceNew - Force create new instance (ignore cache)
 * @returns YooKassaSdk instance
 */
function YooKassa(init, forceNew = false) {
    const cacheKey = init.shop_id;
    const cached = clientCache.get(cacheKey);
    if (!forceNew && cached) {
        return cached;
    }
    const client = new YooKassaSdk(init);
    clientCache.set(cacheKey, client);
    return client;
}
/**
 * Очищает кэш инстансов SDK.
 * Полезно при смене credentials или для освобождения памяти.
 *
 * @param shopId - ID магазина для удаления из кэша. Если не указан, очищается весь кэш.
 */
function clearYooKassaCache(shopId) {
    if (shopId) {
        clientCache.delete(shopId);
    }
    else {
        clientCache.clear();
    }
}
