// @webzaytsev/yookassa-ts-sdk/dist/webhooks/notification.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebhookValidationError = exports.YOOKASSA_IPV6_RANGE = exports.YOOKASSA_IP_RANGES = void 0;
exports.isYooKassaIP = isYooKassaIP;
exports.parseNotification = parseNotification;
exports.parsePaymentNotification = parsePaymentNotification;
exports.parseRefundNotification = parseRefundNotification;
/**
 * IP-адреса и диапазоны, с которых YooKassa отправляет уведомления.
 * Используйте для валидации входящих запросов.
 *
 * @see https://yookassa.ru/developers/using-api/webhooks#ip
 */
exports.YOOKASSA_IP_RANGES = [
    // IPv4 диапазоны
    '185.71.76.0/27',
    '185.71.77.0/27',
    '77.75.153.0/25',
    '77.75.154.128/25',
    // Отдельные IPv4 адреса
    '77.75.156.11',
    '77.75.156.35',
];
/**
 * IPv6 диапазон YooKassa
 * @see https://yookassa.ru/developers/using-api/webhooks#ip
 */
exports.YOOKASSA_IPV6_RANGE = '2a02:5180::/32';
/**
 * Ошибка валидации уведомления
 */
class WebhookValidationError extends Error {
    constructor(message, code) {
        super(message);
        this.code = code;
        this.name = 'WebhookValidationError';
        // Сохраняем оригинальный stack trace
        if (Error.captureStackTrace) {
            Error.captureStackTrace(this, WebhookValidationError);
        }
    }
}
exports.WebhookValidationError = WebhookValidationError;
/**
 * Проверяет, является ли IP-адрес адресом YooKassa.
 *
 * @param ip - IP-адрес для проверки (например, из req.ip или x-forwarded-for)
 * @returns true, если IP принадлежит YooKassa
 *
 * @example
 * ```ts
 * import { isYooKassaIP } from 'yookassa-ts-sdk'
 *
 * app.post('/webhook', (req, res) => {
 *     const clientIP = req.ip || req.headers['x-forwarded-for']
 *     if (!isYooKassaIP(clientIP)) {
 *         return res.status(403).send('Forbidden')
 *     }
 *     // ...
 * })
 * ```
 */
function isYooKassaIP(ip) {
    // Убираем IPv6 prefix если есть (::ffff:192.168.1.1)
    const cleanIP = ip.replace(/^::ffff:/, '');
    // Проверяем IPv6
    if (cleanIP.includes(':')) {
        return isIPv6InRange(cleanIP, exports.YOOKASSA_IPV6_RANGE);
    }
    // Проверяем IPv4
    for (const range of exports.YOOKASSA_IP_RANGES) {
        if (isIPv4Match(cleanIP, range)) {
            return true;
        }
    }
    return false;
}
/**
 * Проверяет, соответствует ли IPv4 адрес диапазону CIDR или отдельному IP
 */
function isIPv4Match(ip, rangeOrIP) {
    // Если это отдельный IP (без маски)
    if (!rangeOrIP.includes('/')) {
        return ip === rangeOrIP;
    }
    // CIDR диапазон
    const [rangeIP, bits] = rangeOrIP.split('/');
    const mask = ~(Math.pow(2, (32 - parseInt(bits, 10))) - 1);
    const ipNum = ipToNumber(ip);
    const rangeNum = ipToNumber(rangeIP);
    if (ipNum === null || rangeNum === null) {
        return false;
    }
    return (ipNum & mask) === (rangeNum & mask);
}
/**
 * Проверяет, входит ли IPv6 адрес в CIDR диапазон
 */
function isIPv6InRange(ip, cidr) {
    const [rangeIP, bits] = cidr.split('/');
    const prefixBits = parseInt(bits, 10);
    const ipParts = expandIPv6(ip);
    const rangeParts = expandIPv6(rangeIP);
    if (!ipParts || !rangeParts) {
        return false;
    }
    // Сравниваем побитово
    let bitsToCompare = prefixBits;
    for (let i = 0; i < 8 && bitsToCompare > 0; i++) {
        const bitsInThisPart = Math.min(16, bitsToCompare);
        const mask = bitsInThisPart === 16 ? 0xffff : (0xffff << (16 - bitsInThisPart)) & 0xffff;
        if ((ipParts[i] & mask) !== (rangeParts[i] & mask)) {
            return false;
        }
        bitsToCompare -= 16;
    }
    return true;
}
/**
 * Разворачивает сокращённый IPv6 адрес в массив из 8 чисел
 */
function expandIPv6(ip) {
    // Обработка :: (сокращение нулей)
    const parts = ip.split(':');
    const emptyIndex = parts.indexOf('');
    if (emptyIndex !== -1) {
        // Убираем пустые строки от ::
        const nonEmpty = parts.filter((p) => p !== '');
        const zerosNeeded = 8 - nonEmpty.length;
        const zeros = Array(zerosNeeded).fill('0');
        parts.splice(emptyIndex, parts.filter((p) => p === '').length, ...zeros);
    }
    if (parts.length !== 8) {
        return null;
    }
    const result = [];
    for (const part of parts) {
        const num = parseInt(part, 16);
        if (Number.isNaN(num) || num < 0 || num > 0xffff) {
            return null;
        }
        result.push(num);
    }
    return result;
}
/**
 * Преобразует IP-адрес в число
 */
function ipToNumber(ip) {
    const parts = ip.split('.');
    if (parts.length !== 4) {
        return null;
    }
    let num = 0;
    for (const part of parts) {
        const octet = parseInt(part, 10);
        if (Number.isNaN(octet) || octet < 0 || octet > 255) {
            return null;
        }
        num = (num << 8) + octet;
    }
    return num >>> 0; // Convert to unsigned
}
/**
 * Парсит и валидирует входящее уведомление от YooKassa.
 *
 * @param body - Тело запроса (req.body)
 * @returns Типизированное уведомление
 * @throws {WebhookValidationError} Если формат уведомления некорректен
 *
 * @example
 * ```ts
 * import { parseNotification } from 'yookassa-ts-sdk'
 *
 * app.post('/webhook', (req, res) => {
 *     try {
 *         const notification = parseNotification(req.body)
 *
 *         switch (notification.event) {
 *             case 'payment.succeeded':
 *                 const payment = notification.object
 *                 console.log('Payment succeeded:', payment.id)
 *                 break
 *             case 'payment.canceled':
 *                 console.log('Payment canceled:', notification.object.id)
 *                 break
 *             case 'refund.succeeded':
 *                 console.log('Refund succeeded:', notification.object.id)
 *                 break
 *         }
 *
 *         res.status(200).send('OK')
 *     } catch (error) {
 *         console.error('Invalid webhook:', error)
 *         res.status(400).send('Bad Request')
 *     }
 * })
 * ```
 */
function parseNotification(body) {
    if (!body || typeof body !== 'object') {
        throw new WebhookValidationError('Notification body must be an object');
    }
    const notification = body;
    if (notification.type !== 'notification') {
        throw new WebhookValidationError(`Invalid notification type: expected 'notification', got '${notification.type}'`);
    }
    if (typeof notification.event !== 'string') {
        throw new WebhookValidationError('Missing or invalid event field');
    }
    if (!notification.object || typeof notification.object !== 'object') {
        throw new WebhookValidationError('Missing or invalid object field');
    }
    return notification;
}
/**
 * Типизированный парсер для уведомлений о платежах.
 *
 * @param body - Тело запроса
 * @returns Уведомление с типизированным объектом платежа
 * @throws {WebhookValidationError} Если событие не относится к платежам
 */
function parsePaymentNotification(body) {
    const notification = parseNotification(body);
    if (!notification.event.startsWith('payment.')) {
        throw new WebhookValidationError(`Expected payment event, got '${notification.event}'`);
    }
    return notification;
}
/**
 * Типизированный парсер для уведомлений о возвратах.
 *
 * @param body - Тело запроса
 * @returns Уведомление с типизированным объектом возврата
 * @throws {WebhookValidationError} Если событие не относится к возвратам
 */
function parseRefundNotification(body) {
    const notification = parseNotification(body);
    if (!notification.event.startsWith('refund.')) {
        throw new WebhookValidationError(`Expected refund event, got '${notification.event}'`);
    }
    return notification;
}
