// @webzaytsev/yookassa-ts-sdk/dist/types/payments/paymentMethod.type.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BankCardSourceEnum = exports.BankCardTypeEnum = exports.PaymentMethodsEnum = void 0;
var PaymentMethodsEnum;
(function (PaymentMethodsEnum) {
    /** Банковская карта или карта МИР */
    PaymentMethodsEnum["bank_card"] = "bank_card";
    /** ЮMoney */
    PaymentMethodsEnum["yoo_money"] = "yoo_money";
    /**
     * QIWI Кошелек
     * @deprecated QIWI Банк лишён лицензии ЦБ РФ 21.02.2024. Способ оплаты не работает.
     */
    PaymentMethodsEnum["qiwi"] = "qiwi";
    /** SberPay */
    PaymentMethodsEnum["sberbank"] = "sberbank";
    /**
     * Альфа-Клик
     * @deprecated Сервис устарел. Рекомендуется использовать другие способы оплаты.
     */
    PaymentMethodsEnum["alfabank"] = "alfabank";
    /** Тинькофф (T-Pay) */
    PaymentMethodsEnum["tinkoff_bank"] = "tinkoff_bank";
    /** СберБанк Бизнес Онлайн */
    PaymentMethodsEnum["b2b_sberbank"] = "b2b_sberbank";
    /** СБП (Система быстрых платежей) */
    PaymentMethodsEnum["sbp"] = "sbp";
    /** Баланс телефона */
    PaymentMethodsEnum["mobile_balance"] = "mobile_balance";
    /** Наличные */
    PaymentMethodsEnum["cash"] = "cash";
    /** Заплатить по частям */
    PaymentMethodsEnum["installments"] = "installments";
    /** "Покупки в кредит" от Сбербанка" */
    PaymentMethodsEnum["sber_loan"] = "sber_loan";
    /** Плати частями (BNPL от СберБанка) */
    PaymentMethodsEnum["sber_bnpl"] = "sber_bnpl";
    /** Alfa Pay */
    PaymentMethodsEnum["alfa_pay"] = "alfa_pay";
    /** Apple Pay (обычно используется через payment_token) */
    PaymentMethodsEnum["apple_pay"] = "apple_pay";
    /** Google Pay (обычно используется через payment_token) */
    PaymentMethodsEnum["google_pay"] = "google_pay";
    /**
     * WebMoney
     * @deprecated WebMoney прекратил работу в РФ в 2022 году.
     */
    PaymentMethodsEnum["webmoney"] = "webmoney";
    /**
     * WeChat Pay
     * @deprecated Специфичный способ оплаты для китайских пользователей.
     */
    PaymentMethodsEnum["wechat"] = "wechat";
})(PaymentMethodsEnum || (exports.PaymentMethodsEnum = PaymentMethodsEnum = {}));
/** Тип банковской карты
 * @see https://yookassa.ru/developers/api#payment_object_payment_method_card_card_type
 */
var BankCardTypeEnum;
(function (BankCardTypeEnum) {
    BankCardTypeEnum["MasterCard"] = "MasterCard";
    BankCardTypeEnum["Visa"] = "Visa";
    BankCardTypeEnum["Mir"] = "Mir";
    BankCardTypeEnum["UnionPay"] = "UnionPay";
    BankCardTypeEnum["JCB"] = "JCB";
    BankCardTypeEnum["AmericanExpress"] = "AmericanExpress";
    BankCardTypeEnum["DinersClub"] = "DinersClub";
    BankCardTypeEnum["DiscoverCard"] = "DiscoverCard";
    BankCardTypeEnum["InstaPayment"] = "InstaPayment";
    BankCardTypeEnum["InstaPaymentTM"] = "InstaPaymentTM";
    BankCardTypeEnum["Laser"] = "Laser";
    BankCardTypeEnum["Dankort"] = "Dankort";
    BankCardTypeEnum["Solo"] = "Solo";
    BankCardTypeEnum["Switch"] = "Switch";
    BankCardTypeEnum["Unknown"] = "Unknown";
})(BankCardTypeEnum || (exports.BankCardTypeEnum = BankCardTypeEnum = {}));
/** Источник данных банковской карты (если использовался Pay-сервис)
 * @see https://yookassa.ru/developers/api#payment_object_payment_method_card_source
 */
var BankCardSourceEnum;
(function (BankCardSourceEnum) {
    /** Apple Pay */
    BankCardSourceEnum["apple_pay"] = "apple_pay";
    /** Google Pay */
    BankCardSourceEnum["google_pay"] = "google_pay";
    /** Mir Pay */
    BankCardSourceEnum["mir_pay"] = "mir_pay";
})(BankCardSourceEnum || (exports.BankCardSourceEnum = BankCardSourceEnum = {}));
